test/unit/integrations/paytm/paytm_module_test.rb

```diff
@@ -0,0 +1,152 @@
+require 'test_helper'
+require 'openssl'
+require 'base64'
+require 'digest'
+require 'securerandom'
+	
+	
+  
+
+class PaytmModuleTest < Test::Unit::TestCase
+  include OffsitePayments::Integrations
+  
+  def new_pg_encrypt(params)
+    if (params.class != Hash) || (params.keys == [])
+      return false
+    end
+    if !params.has_key?(:key)
+      return false
+    end
+    encrypted_data = Hash[]
+    key = params.delete(:key)
+    keys = params.keys
+    aes = OpenSSL::Cipher::Cipher.new("aes-128-cbc")
+    begin
+      keys.each do |k|
+        data = params[k]
+        aes.encrypt
+        aes.key = key
+        aes.iv = '@@@@&&&&####$$$$'
+        encrypted_k = aes.update(k.to_s) + aes.final
+        encrypted_k = Base64.encode64(encrypted_k.to_s)
+        aes.encrypt
+        aes.key = key
+        aes.iv = '@@@@&&&&####$$$$'
+        encrypted_data[encrypted_k] = aes.update(data.to_s) + aes.final
+        encrypted_data[encrypted_k] = Base64.encode64(encrypted_data[encrypted_k])
+      end
+    rescue Exception => e
+      return false
+    end
+    return encrypted_data
+  end
+  
+  
+  def new_pg_encrypt_variable(data, key)
+    aes = OpenSSL::Cipher::Cipher.new("aes-128-cbc")
+    aes.encrypt
+    aes.key = key
+    aes.iv = '@@@@&&&&####$$$$'
+    encrypted_data = nil
+    begin
+      encrypted_data = aes.update(data.to_s) + aes.final
+      encrypted_data = Base64.encode64(encrypted_data)
+    rescue Exception => e
+      return false
+    end
+    return encrypted_data
+  end 
+  
+
+  def new_pg_generate_salt(length)
+    salt = SecureRandom.urlsafe_base64(length*(3.0/4.0))
```

Comment:
Hi Andrew/Alexander,

We have fixed the point raised by your team

Regarding


*salt = SecureRandom.urlsafe_base64(length*(3.0/4.0))**Where does
this 3.0/4.0 come from? Why those numbers?*



*This is because we need 4 digit salt and urlsafe_base64 return n*4/3
digits so to get a salt of 4 digit formula is (4(3/4)) for your reference
kindly check **http://ruby-doc.org/stdlib-1.9.3/libdoc/securerandom/rdoc/SecureRandom.html
<http://ruby-doc.org/stdlib-1.9.3/libdoc/securerandom/rdoc/SecureRandom.html>*


Regarding

*Using a fixed iv is dangerous, should be randomized on each encrypt and
included at the beginning of the encrypted
data. https://cwe.mitre.org/data/definitions/329.html
<https://cwe.mitre.org/data/definitions/329.html>*


*as per our encryption logic we need a fixed iv, we can't randomize it.*


Kindly Check,

Regards,

On Tue, Jun 7, 2016 at 12:42 PM, Mohit Aggarwal <mohit.aggarwal@paytm.com>
wrote:

> Hi ,
>
> Thanks for you inputs, we are working on this will get back to you shortly
>
> Regards,
>
> On Mon, Jun 6, 2016 at 10:00 PM, Alexander Profeit <
> notifications@github.com> wrote:
>
>> In test/unit/integrations/paytm/paytm_module_test.rb
>> <https://github.com/activemerchant/offsite_payments/pull/200#discussion_r65922759>
>> :
>>
>> > +    aes.encrypt
>> > +    aes.key = key
>> > +    aes.iv = '@@@@&&&&####$$$$'
>> > +    encrypted_data = nil
>> > +    begin
>> > +      encrypted_data = aes.update(data.to_s) + aes.final
>> > +      encrypted_data = Base64.encode64(encrypted_data)
>> > +    rescue Exception => e
>> > +      return false
>> > +    end
>> > +    return encrypted_data
>> > +  end
>> > +
>> > +
>> > +  def new_pg_generate_salt(length)
>> > +    salt = SecureRandom.urlsafe_base64(length*(3.0/4.0))
>>
>> Where does this 3.0/4.0 come from? Why those numbers?
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/activemerchant/offsite_payments/pull/200/files/59723b4bb7a66b4abdf17fbbcdc5a20ec39afcf5#r65922759>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/AL1PqNfHbYLhUpgcSPnYKZhb1YXb8ROuks5qJEtCgaJpZM4IQnF4>
>> .
>>
>
>
>
> --
> *Mohit Aggarwal*
> Senior Software Engineer
> +91 8826036707
> [image: Inline image 1]
> www.Paytm.com
>



-- 
*Mohit Aggarwal*
Senior Software Engineer
+91 8826036707
[image: Inline image 1]
www.Paytm.com

-- 


Disclaimer :-

This email and any files transmitted with it are confidential and 
intended solely for the use of the individual or entity to which they are 
addressed. If you are not the intended recipient, you should not 
disseminate, distribute or copy this e-mail. Please notify the sender 
immediately by e-mail and destroy all copies of this message and any 
attachments. Any views or opinions presented in this email are solely those 
of the author and do not necessarily represent those of the company. 

Warning: Although the company has taken reasonable precautions to ensure no 
viruses are present in this email, the company cannot accept responsibility 
for any loss or damage arising from the use of this email or attachments.